py/desitarget/mock/build.py

```diff
@@ -4,1488 +4,439 @@
-                if 'LYA' in params['sources'][source_name].keys():
-                    density_lya = params['sources'][source_name]['LYA']['density']
-                    zcut = params['sources'][source_name]['LYA']['zcut']
-                    tracer = truth['TRUEZ'] < zcut
-                    lya = truth['TRUEZ'] >= zcut
-                    if len(tracer) > 0:
-                        log.info('Downsampling tracer {}s to desired target density of {} targets/deg2.'.format(target_name, density))
-                        Selection.density_select(targets, truth, source_name=source_name, target_name=target_name,
-                                                 density=density, subset=tracer)
-                        print()
-                    if len(lya) > 0:
-                        log.info('Downsampling Lya {}s to desired target density of {} targets/deg2.'.format(target_name, density_lya))
-                        Selection.density_select(targets, truth, source_name=source_name, target_name=target_name,
-                                                 density=density_lya, subset=lya)
-
-                else:
-                    Selection.density_select(targets, truth, source_name=source_name,
-                                             target_name=target_name, density=density)
-                    
-            else:
-                Selection.density_select(targets, truth, source_name=source_name,
-                                         target_name=target_name, density=density)            
-
-            keep = np.where(targets['DESI_TARGET'] != 0)[0]
-            #keep = np.where( np.any( ((targets['DESI_TARGET'] != 0), (truth['CONTAM_TARGET'] != 0)), axis=0) )[0]
-
-            if len(keep) == 0:
-                log.warning('All {} targets rejected!'.format(target_name))
-            else:
-                targets = targets[keep]
-                truth = truth[keep]
-                if target_name.upper() != 'SKY':
-                    trueflux = trueflux[keep, :]
-
-        #rewrite the target ID to be sure that it is unique in this brick
-        if len(targets):
-            new_pixel_ID = current_pixel_ID + len(targets)
-            targets['BRICK_OBJID'][:]= np.arange(current_pixel_ID,new_pixel_ID)
-            current_pixel_ID = new_pixel_ID
-            
-        if target_name.upper() == 'SKY':
-            skytruth = truth.copy()
-            skytargets = targets.copy()
+    # Get the requested target density, if any.
+    ntarget = len(data['RA'])
+    if 'density' in params.keys():
+        density = params['density']
+    else:
+        density = ntarget / areaperpixel
+
+    # Assign sources to healpix chunks.
+    healpix_chunk = radec2pix(nside_chunk, data['RA'], data['DEC'])
+
+    indxperchunk, ntargetperchunk = list(), list()
+    for pixchunk in set(healpix_chunk):
+        indx = np.where( np.in1d(healpix_chunk, pixchunk)*1 )[0]
+        indxperchunk.append(indx)
+
+        if coeff:
```

Comment:
How are Poisson + LSS fluctuations taken into account? Aren't those erased by setting each pixel to a fixed number density?

Reply:
@forero Thank you for this and the previous question (on BGS) regarding how density fluctuations are being handled in this refactor. First, let me say that when I sat down to refactor the code from the brick-based model to a healpixel-based model (which utilized @geordie666's more recent "summary" statistics of the imaging data in #217) I didn't fully understand the fluctuation model that you, Adam, and Stephen had implemented.  However, after speaking with Stephen off-list, I now have a better idea of how you were down-sampling the various mocks to the desired target density while still preserving the imprint of large-scale structure.  

In this refactor I am definitely *not* doing the down-sampling correctly.  However, I'd like to propose -- and I believe that @sbailey agreed -- that we should get this merged in and I'll work on "fixing" the fluctuations model.

Regarding accounting for the imaging systematics:  developing a proper model is something the Clustering WG is (should be?) already doing.  I will reach out to them to see if someone from that group can contribute code so our simulated target densities are as realistic as possible.

Reply:
@moustakas I agree. This is good enough to merge now and implement a fix later.